### ISSUE TITLE
SharedBitmap class to allow re-rendering of static bits of UX throughout the app 

### ISSUE THREAD

**Issue by Duckers** [fusetools/fuselibs-public#172]
## Background
When profiling apps to improve performance, a recurring theme is the time it takes to render visual elements for the first time. In particular when effects like `<DropShadow>` and `<Mask>` is used on tiny elements like icons, allocation of framebuffers and render-to-texture is required for each instance of the icon. If the icon is e.g. repeated for each item in a list, it can add up to hundreds of milliseconds of latency in the middle of transitions.

Of course, the optimal solution here would be to pre-bake the icons with shadows in the image file. However, this is very inconvenient. UX Markup is excellent for describing a "content pipeline" for assets, where final versions can be rendered out after the screen size, density, locale and other runtime properties are available. It would be a shame not to be able to take advantage of that.

## Proposal
Proposing a new class `SharedBitmap`, which is a `LayoutControl` which will be the recommended base class for UX classes that represent small and immutable fixed-size assets, such as icons, buttons. 

The content of `SharedBitmap` is implicitly a `ux:Template` and not instantiated by default. When `SharedBitmap` is used as base class for an `ux:Class`, only the first instance of the class to be rooted actually instantiates the content and renders the content - to a bitmap, which is then shared globally with other instances of the same class. Subsequent instances of the class simply blit out the bitmap and doesn't actually instantiate or render anything.

## Example use

Here is a simple example of an icon with a drop shadow, as a shared bitmap:

```
<SharedBitmap ux:Class="MyApp.FooIcon">
   <Image Source="someMultiDensitySource">
       <DropShadow />
   </Image>
</SharedBitmap>
```

If we now use `<MyApp.FooIcon />` for every card in a feed, for example, only one instance will ever be created and rendered.

## Implementation details

The content of a `SharedBitmap` can be arbitrary and even contain dynamic data bindings that change over time. However, the contract is that it should be possible to assume that every instance of the shared bitmap has the same visual appearance. Each instance of the `SharedBitmap` can have arbitrary transform and opacity without violating this contract.

In order to be able to create the shared cache, the class needs some unique identifier. The proposal here is to have a `ClassId` string constructor argument (ux-dependency) on `SharedBitmap` that is used to look up some global cache, plus an UX compiler feature that automatically fills in this dependency witht the class name as a string when `SharedBitmap` is  the base class in a `ux:Class`. This ensures clean syntax in the base class case, and compile-time enforcing the user to enter an ID when its not the base class.

## Potential future enhancements
- Cache the bitmap to file on disk, and reuse across runs of the app. Will only render the first time the app runs.

**Comment by mortoray** [comment]
Please show an example that would involve an image with a URL coming from JS.

[15:47] 
In particular inside an each — I’m thinking about a flag like system where each item may have a similar set of visual flags applied.

**Comment by mortoray** [comment]
It can't just be the first rooted item that becomes master. There needs to be a fallback to allow another one to takeover if the master item should become unrooted.

**Comment by Duckers** [comment]
The contract is that the visual appearance is always the same for all instances, so as long as there is at least one rooted item, one of them is the master that updates the shared cache,

**Comment by mortoray** [comment]
This is going to come with some complex warnings in the docs about how to use this in layout. The shared class cannot have any part of its layout based on the parent's size. In practice this means that the user will have to provide an explicit `Width` and `Height` or provide an `Alignment` which includes both axes (to be a naturally sized element).

**Comment by Duckers** [comment]
I think `SharedBitmap` should always compact in both directions.

**Comment by eksperts** [comment]
I like the idea because of the performance improvements it brings, but for an end-user this means only one more thing to remember. Meaning, we need to be very explicit about the gains so that it's an obvious choice.

Reading the concerns, something like this would still work, and be cached, right?
```
<SharedBitmap ux:Class="MyApp.FooIcon" SomeBool="false">
   <bool ux:Property="SomeBool" />
   <WhileTrue Value="{ReadProperty SomeBool}">
      <Change theImage.Source="someOtherMultiDensitySource" />
   </WhileTrue>
   <Image ux:Name="theImage" Source="someMultiDensitySource">
       <DropShadow />
   </Image>
</SharedBitmap>
```

And finally, I don't agree with this statement: "`SharedBitmap` should always compact in both directions."
Because semantically I think of `SharedBitmap` as an `Image`, I believe it should act just like one - which AFAIK is _occupy all available space_, if no explicit dimensions or alignments are set.

**Comment by Duckers** [comment]
>Reading the concerns, something like this would still work, and be cached, right?

Yes and no. It will work, but if you have different values for `SomeBool` on different instances, it will not work as you expect: *The contract is that the visual appearance is always the same for all instances*

> semantically I think of SharedBitmap as an Image, I believe it should act just like one - which AFAIK is occupy all available space,

This would require the image to take up exactly the same amount of space wherever used in order to be useful - unlikely in the stretching case.

Note that this feature is predominantly meant for caching icon buttons, which rarely stretch.

**Comment by eksperts** [comment]
Right. It's definitely useful if it brings significant performance boost, but then we need to communicate the exact amount of that impact.

And maybe the name should be something more explicit, to highlight the exact proposed use case; `SharedBitmap` seems to be somewhat obscure, while we're trying to solve a very particular challenge.

**Comment by yupferris** [comment]
> And maybe the name should be something more explicit, to highlight the exact proposed use case; SharedBitmap seems to be somewhat obscure, while we're trying to solve a very particular challenge.

These were my initial thoughts as well; you beat me to it :)

**Comment by yupferris** [comment]
Overall I like the idea, but I have a hard time visualizing some of the more obscure layout consequences it may have (and the aforementioned naming issue, but that's a more minor thing).

**Comment by Duckers** [comment]
Naming things is always difficult. Suggestions are welcome.

**Comment by Duckers** [comment]
Suggestions:
* `StaticAsset`

**Comment by yupferris** [comment]
> Naming things is always difficult. Suggestions are welcome.
Indeed! Right off the bat I like that `SharedBitmap` communicates its static nature. Perhaps `StaticAsset` might be a better name for a base class, to describe what it is more clearly semantically? Avoiding `Bitmap` (or `Image` etc) is probably a good idea to avoid the mental association with `Image` and how it plays with layout, and the `Shared` part, while alluding to the fact that the cached rendering of the element is shared between all instances, has a hint of exposing the underlying implementation.

I think `StaticAsset` clearly communicates that you're describing a class that represents a visual element that will never change, and the fact that the renderer shares a cached rendering between all instances is more of a detail of how `StaticAsset`s are rendered. Further, `ClassId` can be renamed to `AssetId`, which also reads quite nicely in the same context and makes it a bit more explicit why that field exists.

**Comment by yupferris** [comment]
Note that that last comment was in the process of being written while `StaticAsset` was suggested :)

**Comment by Duckers** [comment]
Indeed.

`StaticAsset` also sounds better in the case where the resulting bitmap is cached to file between app runs. It also more intuitively discourages using properties to configure it.

**Comment by eksperts** [comment]
One more suggestion: `FixedSizeSomething` because this thing only does its job when it has _fixed size_. Replace `Something` with something else. I like the `StaticAsset` much better though.

**Comment by yupferris** [comment]
> StaticAsset also sounds better in the case where the resulting bitmap is cached to file between app runs. It also more intuitively discourages using properties to configure it.

yes, exactly.

**Comment by mortoray** [comment]
I don't think this is valuable unless you can use properties to configure it. Even in the case of icons it'll be the standard case to create a common base class and then instantiate them with properties.

```
<MyIcon File="whatever.png"/>
<MyIcon FIle="http://where.ever/item.png"/>
```

Or even using numbers:

```
<MyCountMarker Count="5"/>
<FontAwesome Code="1234"/>
```

With the marker, the most value would be if there is indeed a dynamic Id, and switching the dynamic values of the item could switch the shared resource.

```
<MyCountMarker Count="{count}"/>
```

This would end up storing several images depending on the count.  Or consider a common tag case:

```
<Each Items="{items}">
   <MyTag Value="{tag}"/>
```

You may have only 4-5 tags applied to several hundred items.

Not allowing properties would significanlty stunt the usefulness of this feature and thus not likely be helpful to many users.

**Comment by mortoray** [comment]
The sizing of this item doens't need to be fixed, it just needs to be static. You can either specifiy an explicit `Witdth/Height` in points/pixels values, or you allow it to shrink to it's natural size. We definitely need the natural size option.

**Comment by Duckers** [comment]
I think `Asset` is a good name for this class, actually. It won't neccessarily be "static" anymore. The commonality of the class is that it is an asset which visual representation can be reused throughout the app, and potentially even cached on disk.

**Comment by Duckers** [comment]
Got a first test implementation in the above PR - the basics seem to work fine. Here is a list of labels rendered to a shared buffer:

![image](https://user-images.githubusercontent.com/664425/27879735-dcf98e4c-61c2-11e7-8c60-fee902b1ad77.png)

This is the code:

```<App>
    <Asset ux:Class="MyIcon">
        <Panel>
            <Text>Yolo!</Text>
            <DropShadow Color="#f00" />
        </Panel>
    </Asset>

    <StackPanel>
        <Each Count="10">
            <MyIcon ux:Name="t" Margin="10" Opacity="alternate(index(t),1) ? 0.84 : 1" />
        </Each>
    </StackPanel>
</App>```

**Comment by Duckers** [comment]
@mortoray I have a problem with the `FreezeDrawable` - it doesn't seem to respect rotations.

If I do:

```
            <MyIcon ux:Name="t" Margin="10" Opacity="alternate(index(t),1) ? 0.84 : 1">
                <Rotation Degrees="20" />
            </MyIcon>
```
The visual result isn't as expected, it simply pushes the list slightly down, doesn't rotate the elements.

![image](https://user-images.githubusercontent.com/664425/27879915-81121f26-61c3-11e7-8802-c8a1866341aa.png)

I fear this also affects the PageFreeze feature. Any ideas? Feel free to push to the PR.

Also, i noticed during this that the RenderBounds (and hence the cached rectangles) are way bigger than I would expect (like they are padded with 20 points or something). We've also observed this randomly in some of the stuff @yupferris  is working on. Is this known/expected? Its seems like anything that uses the caching is affected by this, rendering much more pixels than necessary.

**Comment by mortoray** [comment]
I don't recall if I tested transforms on the freeze feature. I have several transitions that involve rotations, but they may not have been involved in that test. It's probably a defect.

Effects will add a lot to the bounds of the objects. Rotations will also add to the bounds, but it's only intended to be the edges of the rotated rectangle.

**Comment by Duckers** [comment]
@yupferris : In the cases where you observed bloated render bounds - can you please veryfiy that you have effects in play there?

**Comment by kusma** [comment]
I've definately seen bloated renderbounds without effects; Text, depending on the settings seems to sometimes report overly conservative render-bounds. I suspect this differs a bit from text-renderer to text-renderer, though.

**Comment by yupferris** [comment]
The bloated text rects I'm seeing don't appear to be affected by effects; I'm seeing bloated rects on elements with and without effects (`DropShadow` in particular). For example, here are a bunch of basic `Text` elements that appear to have large draw rects:
![image](https://user-images.githubusercontent.com/3166056/27907317-7acfb8ba-6247-11e7-8fa7-9410c76d652b.png)

These buttons in the same app have a `DropShadow` applied, and the rects appear to be the same size roughly as the `Text` elements above (assuming the rects extend the same amount below the screen as they do above the buttons):
![image](https://user-images.githubusercontent.com/3166056/27907359-9e0720fc-6247-11e7-80cb-43e2ad7d6f95.png)

**Comment by yupferris** [comment]
I don't currently have a simpler case as I'm only capturing rects from the cache to start with, but this should be simple enough to illustrate the large bounds don't appear to be due to the use of effects.

**Comment by yupferris** [comment]
It appears when using harfbuzz our text rendering measures _very_ conservatively. I've tinted the cache rects to show this (to make sure it wasn't my system borking up):

![image](https://user-images.githubusercontent.com/3166056/27908163-cc5fcdd4-624a-11e7-9f1b-f2311efbdad1.png)

on Android the rects are much tighter except when using `-Duse_harfbuzz`, where it looks the same as .net.

**Comment by Duckers** [comment]
I've deleted the post related to bloated render bounds and spun out this issue instead: https://github.com/fusetools/fuselibs-public/issues/176
